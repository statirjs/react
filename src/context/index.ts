import { createContext } from 'react';
import { Listener } from '@statirjs/core';
import * as S from '../typing/internal';

const context: S.ContextValue = {
  store: {
    state: {},
    dispatch: {},
    listeners: [],
    subscribe(listener: Listener) {
      this.listeners.push(listener);
    }
  }
};

export const Context = createContext(context);
