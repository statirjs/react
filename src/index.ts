export { Provider } from './provider';
export { Context } from './context';
export { useDispath, useSelector } from './hooks';
