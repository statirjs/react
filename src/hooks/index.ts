import { useMemo, useCallback, useContext } from 'react';
import { Context } from '../context';
import * as S from '../typing/internal';

export function useSelector<T>(selector: S.StateSelector<T>): T {
  const { store } = useContext(Context);

  return useMemo(() => selector(store.state), [store, selector]);
}

export function useDispath<T extends S.Action>(
  selector: S.DispatchSelector<T>
): T {
  const { store, reload } = useContext(Context);

  const action = useMemo(() => selector(store.dispatch), [store, selector]);

  return useCallback(
    (...rest: S.Payload[]) => {
      action(...rest);
      reload();
    },
    [action, reload]
  ) as T;
}
