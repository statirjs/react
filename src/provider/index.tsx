import React, { useMemo, useState, useCallback, memo } from 'react';
import { Context } from '../context';
import * as S from '../typing/internal';

const INIT_RELOADER = {};

function ProviderComponent(props: S.ProviderProps) {
  const { store, children = null } = props;

  const [reloader, setReload] = useState(INIT_RELOADER);

  const reload = useCallback(() => setReload({ ...reloader }), [
    reloader,
    setReload
  ]);

  const value = useMemo(
    () => ({
      store,
      reload
    }),
    [store, reload, reloader]
  );

  return <Context.Provider value={value}>{children}</Context.Provider>;
}

export const Provider = memo(ProviderComponent);
