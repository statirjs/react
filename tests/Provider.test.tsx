import React from 'react';
import renderer from 'react-test-renderer';
import { createForme, initStore } from '@statirjs/core';
import { Provider } from '../src';

const counter = createForme(
  {
    count: 0
  },
  () => ({
    actions: {
      increment(state) {
        return {
          ...state,
          count: state.count + 1
        };
      }
    }
  })
);

const store = initStore({
  forms: {
    counter
  }
});

describe('Test Provider', () => {
  test('snapshot', () => {
    const component = renderer.create(<Provider store={store} />);

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
