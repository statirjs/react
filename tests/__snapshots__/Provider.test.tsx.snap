// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Provider snapshot 1`] = `null`;
